snuba/datasets/outcomes.py

```diff
@@ -0,0 +1,164 @@
+            ('timestamp', DateTime()),
+            ('outcome', UInt(8)),
+            ('reason', Nullable(String())),
+            ('event_id', Nullable(String())),
+        ])
+
+    def get_local_engine(self):
+        return """
+            ReplicatedMergeTree('/clickhouse/tables/{shard}/%(name)s', '{replica}')
+            PARTITION BY %(partition_by)s
+            ORDER BY %(order_by)s;""" % {
+                'name': self.LOCAL_TABLE,
+                'order_by': self.ORDER_BY,
+                'partition_by': self.PARTITION_BY,
+            }
+
+
+class DevOutcomesTableSchema(OutcomesTableSchema):
+    def __init__(self, *args, **kwargs):
+        super(DevOutcomesTableSchema, self).__init__(*args, **kwargs)
+
+        self.LOCAL_TABLE = 'dev_outcomes'
+        self.QUERY_TABLE = self.LOCAL_TABLE
+        self.CAN_DROP = True
+
+    def get_local_engine(self):
+        return """
+            MergeTree()
+            PARTITION BY %(partition_by)s
+            ORDER BY %(order_by)s;""" % {
+                'name': self.LOCAL_TABLE,
+                'order_by': self.ORDER_BY,
+                'partition_by': self.PARTITION_BY,
+            }
+
+class TestOutcomesTableSchema(DevOutcomesTableSchema):
+    def __init__(self, *args, **kwargs):
+        super(TestOutcomesTableSchema, self).__init__(*args, **kwargs)
+
+        self.LOCAL_TABLE = 'test_outcomes'
+        self.QUERY_TABLE = self.LOCAL_TABLE
+        self.CAN_DROP = True
+
+class OutcomesProcessor(Processor):
+    KAFKA_MESSAGE_SCHEMA = {
+        'type': 'object',
+        'properties': {
+            'org_id': {'type': 'number'},
+            'project_id': {'type': 'number'},
+            'key_id': {'type': ['number', 'null']},
+            'timestamp': {
+                'type': 'string',
+                'format': 'date-time',
+            },
+            'outcome': {
+                'type': ['string', 'number'],
+                'min': 0,
+                'max': 255
+            },
+            'reason': {'type': ['string', 'null']},
+            'event_id': {'type': ['string', 'null']},
+        },
+        'required': [
+            'timestamp',
+            'org_id',
+            'project_id',
+            'outcome'
+        ],
+        'additionalProperties': False,
+    }
+
+    def __init__(self, SCHEMA):
+        super(OutcomesProcessor, self).__init__(SCHEMA)
+
+        self.MESSAGE_TOPIC = 'outcomes'
+        self.MESSAGE_CONSUMER_GROUP = 'outcome-consumers'
+        self.REPLACEMENTS_TOPIC = None
+        self.REPLACEMENTS_CONSUMER_GROUP = None
+        self.COMMIT_LOG_TOPIC = None
+
+    def validate_message(self, message):
+        try:
+            schemas.validate(message, self.KAFKA_MESSAGE_SCHEMA, False)
+        except (ValueError, jsonschema.ValidationError) as e:
+            # TODO separate message type, version etc? or just coalesce all these exception types
+            raise InvalidMessage("Invalid message: {}".format(e))
```

Comment:
Could you have unicode in the exception message here?

Reply:
technically I don't think so, as jsonschema.ValidationErrors come back with a `repr()`d version of the payload, which is a non-unicode string, but I will add the u there just in case there is something that could throw a ValueError with a unicode string